platform/security/idp/security-idp-plugin/security-idp-plugin-requestsubjectvalidator/src/main/java/org/codice/ddf/security/idp/plugin/requestsubjectvalidator/RequestSubjectValidatorPresignPlugin.java

```diff
@@ -0,0 +1,103 @@
+import org.opensaml.saml.common.SAMLRuntimeException;
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.NameID;
+import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.Subject;
+
+/**
+ * This plugin is responsible for verifying that the {@code Subject} of the received {@code
+ * AuthnRequest} refers to the same principle as the {@code Subject} of the {@code Response}.
+ */
+public class RequestSubjectValidatorPresignPlugin implements SamlPresignPlugin {
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<Binding> supportedBindings) {
+
+    Subject authnRequestSubject = authnRequest.getSubject();
+    if (authnRequestSubject == null) {
+      return;
+    }
+
+    if (!authnRequestSubject.getSubjectConfirmations().isEmpty()) {
+      throw new SAMLRuntimeException(
+          "The AuthnRequest contained a Subject with 1 or more SubjectConfirmations.");
+    }
+
+    NameID authnRequestNameId = authnRequestSubject.getNameID();
+    if (authnRequestNameId == null) {
+      return;
+    }
+
+    String authnRequestId = authnRequestNameId.getValue();
+    if (StringUtils.isBlank(authnRequestId)) {
+      return;
+    }
+
+    if (idsFromResponseAsStream(response).noneMatch(authnRequestId::equals)) {
+      throw new SAMLRuntimeException(
+          String.format(
+              "The AuthnRequest's Subject (with the NameID of %s) could not be matched against any "
+                  + "identifiers in the resulting Response (with the primary NameID of %s).",
+              authnRequestId, nameIdsFromResponseAsStream(response).findFirst().orElse(null)));
+    }
+  }
+
+  /** Returns all the identifiers that were returned on the response, i.e. username, email, ect. */
+  private Stream<String> idsFromResponseAsStream(Response response) {
+    return Stream.concat(
+        nameIdsFromResponseAsStream(response), attributeIdsFromResponseAsStream(response));
+  }
+
+  private Stream<String> nameIdsFromResponseAsStream(Response response) {
+    return response
+        .getAssertions()
+        .stream()
+        .map(Assertion::getSubject)
+        .map(Subject::getNameID)
+        .filter(Objects::nonNull)
+        .map(NameID::getValue)
+        .filter(Objects::nonNull)
+        .filter(StringUtils::isNotBlank);
+  }
+
+  private Stream<String> attributeIdsFromResponseAsStream(Response response) {
+    return response
+        .getAssertions()
+        .stream()
+        .flatMap(assertion -> assertion.getAttributeStatements().stream())
+        .flatMap(statement -> statement.getAttributes().stream())
+        .flatMap(attribute -> attribute.getAttributeValues().stream())
+        .filter(attributeValue -> attributeValue instanceof XSString)
+        .map(attributeValue -> (XSString) attributeValue)
```

Comment:
✏️ `.map(XSString::cast)`

Reply:
Looks like `XSString` does not have a `cast` method. Maybe because it is an interface?

Reply:
Sorry ... `XSString.class::isInstance()`